pkg/analyze/data_source_test.go

```diff
@@ -0,0 +1,265 @@
+				})
+			}
+
+			ctx := context.Background()
+			listFunc := func(ctx context.Context, options metav1.ListOptions) (runtime.Object, error) {
+				return &metav1.List{
+					TypeMeta: metav1.TypeMeta{},
+					ListMeta: metav1.ListMeta{},
+					Items:    objectsAsRawExtension,
+				}, nil
+			}
+
+			lister, err := BuildLister[dummyLister](ctx, test.ListerFactory, listFunc)
+
+			if err != nil {
+				t.Fatalf("BuildLister returned an error: %v", err)
+			}
+			if reflect.TypeOf(lister) != reflect.TypeFor[dummyLister]() {
+				t.Fatalf("expected '%v' got: '%v'", reflect.TypeFor[dummyLister](), reflect.TypeOf(lister))
+			}
+
+			actualObjects, err := lister.List()
+			if err != nil {
+				t.Fatalf("dummyLister.List returned an error: %v", err)
+			}
+
+			assertExpectedContainsPermutatedActual(t, test.Objects, actualObjects)
+		})
+	}
+}
+
+func compareSingle[T any, U *T](
+	t *testing.T,
+	all []runtime.Object,
+	listFunc func(selector labels.Selector) (ret []*T, err error),
+) {
+	actualResources, err := listFunc(labels.Everything())
+	if err != nil {
+		t.Fatalf("%s lister returned an error: %v", reflect.TypeFor[U]().Name(), err)
+	}
+
+	var expectedResources []U
+	for _, res := range all {
+		switch v := res.(type) {
+		case U:
+			expectedResources = append(expectedResources, v)
+		}
+	}
+
+	assertExpectedContainsPermutatedActual(t, expectedResources, actualResources)
+}
+
+func TestNewDataSourceFromClients(t *testing.T) {
+	newDataSourceTests := map[string]struct {
+		kubernetesObjects []runtime.Object
+		scyllaObjects     []runtime.Object
+	}{
+		"no objects": {},
+		"some listers empty": {
+			kubernetesObjects: []runtime.Object{
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Pod1",
+						Namespace: "NS1",
+					},
+				},
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Pod2",
+						Namespace: "NS1",
+					},
+				},
+				&v1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Svc1",
+						Namespace: "NS1",
+					},
+				},
+			},
+			scyllaObjects: []runtime.Object{
+				&scyllav1.ScyllaCluster{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "ScyllaCluster1",
+						Namespace: "NS1",
+					},
+				},
+			},
+		},
+		"all objects": {
+			kubernetesObjects: []runtime.Object{
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Pod1",
+						Namespace: "NS1",
+					},
+				},
+				&v1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Svc1",
+						Namespace: "NS1",
+					},
+				},
+				&v1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Secret1",
+						Namespace: "NS2",
+					},
+				},
+				&v1.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "ConfigMap1",
+						Namespace: "NS1",
+					},
+				},
+			},
+			scyllaObjects: []runtime.Object{
+				&scyllav1.ScyllaCluster{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "ScyllaCluster1",
+						Namespace: "NS1",
+					},
+				},
+			},
+		},
+	}
```

Comment:
it seems like having a test for every supported lister type would be beneficial. It's a lot of bloat, but since there's a lot of repetition in the constructor it's easy to miss a bug there (like a non-wired lister). I think one empty and non-empty case per lister should be enough? You can have a func returning a copy of a set of objects which you can tweak per test, to make it cleaner

Reply:
done, it's a bit ugly but I don't have a better idea how to do this

Reply:
this is what I had in mind, seems pretty much aligned with Maciek's comment as well
```
func TestNewDataSourceFromClients(t *testing.T) {
	t.Parallel()

	tt := []struct {
		name              string
		kubernetesObjects []runtime.Object
		scyllaObjects     []runtime.Object
		listerFunc        func(*DataSource) ([]runtime.Object, error)
		expectedObjects   []runtime.Object
		expectedErr       error
	}{
		{
			name: "empty pod list",
			kubernetesObjects: func() []runtime.Object {
				return slices.FilterOut(newBasicKubernetesObjects(), func(obj runtime.Object) bool {
					_, ok := obj.(*corev1.Pod)
					return ok
				})
			}(),
			scyllaObjects: newBasicScyllaObjects(),
			listerFunc: func(ds *DataSource) ([]runtime.Object, error) {
				pods, err := ds.PodLister.List(labels.Everything())
				if err != nil {
					return nil, fmt.Errorf("can't list pods: %w", err)
				}

				return slices.ConvertSlice(pods, convertToRuntimeObject), nil
			},
			expectedObjects: []runtime.Object{},
			expectedErr:     nil,
		},
		{
			name:              "nonempty pod list",
			kubernetesObjects: newBasicKubernetesObjects(),
			scyllaObjects:     newBasicScyllaObjects(),
			listerFunc: func(ds *DataSource) ([]runtime.Object, error) {
				pods, err := ds.PodLister.List(labels.Everything())
				if err != nil {
					return nil, fmt.Errorf("can't list pods: %w", err)
				}

				return slices.ConvertSlice(pods, convertToRuntimeObject), nil
			},
			expectedObjects: []runtime.Object{
				&corev1.Pod{
					ObjectMeta: metav1.ObjectMeta{
						Name:      "pod1",
						Namespace: "test",
					},
				},
				&corev1.Pod{
					ObjectMeta: metav1.ObjectMeta{
						Name:      "pod2",
						Namespace: "test",
					},
				},
			},
			expectedErr: nil,
		},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			ctx, cancel := context.WithCancel(context.Background())
			defer cancel()

			fakeClient := fake.NewSimpleClientset(tc.kubernetesObjects...)
			scyllaFakeClient := scyllafake.NewSimpleClientset(tc.scyllaObjects...)

			dataSource, err := NewDataSourceFromClients(ctx, fakeClient, scyllaFakeClient)
			if !reflect.DeepEqual(err, tc.expectedErr) {
				t.Fatalf("expected error: %v, got: %v", tc.expectedErr, err)
			}

			objects, err := tc.listerFunc(dataSource)
			if err != nil {
				t.Fatalf("unexpected error from lister func: %v", err)
			}

			if !equality.Semantic.DeepEqual(objects, tc.expectedObjects) {
				t.Errorf("expected and actual objects differ: %s", cmp.Diff(tc.expectedObjects, objects))
			}
		})
	}
}

func convertToRuntimeObject[T runtime.Object](v T) runtime.Object { return runtime.Object(v) }

func newBasicKubernetesObjects() []runtime.Object {
	return []runtime.Object{
		&corev1.Pod{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "pod1",
				Namespace: "test",
			},
		},
		&corev1.Pod{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "pod2",
				Namespace: "test",
			},
		},
	}
}

func newBasicScyllaObjects() []runtime.Object {
	return []runtime.Object{
		&scyllav1alpha1.ScyllaDBDatacenter{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "scylladbdatacenter1",
				Namespace: "test",
			},
		}, &scyllav1alpha1.ScyllaDBDatacenter{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "scylladbdatacenter2",
				Namespace: "test",
			},
		},
	}
}
```

overall being explicit and verbose in unit tests is fine and desired, readability is much more important here than avoiding repetitions; other than actually testing the code, it also let's people see what the desired behaviour is

Reply:
done